core/src/main/scala/kafka/admin/ConfigCommand.scala

```diff
@@ -114,9 +116,21 @@ object ConfigCommand extends Config {
 
     if (entityType == ConfigType.User)
       preProcessScramCredentials(configsToBeAdded)
-    if (entityType == ConfigType.Broker) {
-      require(configsToBeAdded.asScala.keySet.forall(BrokerConfigsUpdatableUsingZooKeeper.contains),
-        s"--bootstrap-server option must be specified to update broker configs $configsToBeAdded")
+    else if (entityType == ConfigType.Broker) {
+      // Replication quota configs may be updated using ZK at any time. Other dynamic broker configs
+      // may be updated using ZooKeeper only if the corresponding broker is not running. Dynamic broker
+      // configs at cluster-default level may be configured using ZK only if there are no brokers running.
```

Comment:
This seems a little annoying from a user perspective. If we have to open the door for changing configs directly through zk, do we get much benefit from restricting its usage?

Reply:
@hachikuji Thanks for the review. For initial configuration of brokers, we don't do much validation. This feels reasonable since initial config from server.properties is only validated when broker starts up (broker fails to start with invalid config). The validation is slightly better with configs persisted in ZK since we can check the types etc. With dynamic configuration of a running broker, we do a lot more validation. For example, for keystore update, we can check that the file exists on the broker and is a valid store, but that can only be verified by the broker itself. For inter-broker keystore update, we can validate keystore against truststore. We could allow updating using ZK at any time and have the broker print out an error and not apply an invalid config. But since we are able to validate using AdminClient and the tool gives immediate feedback for invalid config, I thought it was better to restrict this for bootstrapping use only.